pkg/ha/filter.go

```diff
@@ -49,42 +51,125 @@ func (h *Filter) Process(_ *http.Request, wr *prompb.WriteRequest) error {
 
 	minT := model.Time(minTUnix).Time()
 	maxT := model.Time(maxTUnix).Time()
-	allowInsert, acceptedMinT, err := h.service.CheckLease(minT, maxT, clusterName, replicaName)
+	allowInsert, leaseStart, err := h.service.CheckLease(minT, maxT, clusterName, replicaName)
 	if err != nil {
 		return fmt.Errorf("could not check ha lease: %#v", err)
 	}
-	if !allowInsert {
-		log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
-		wr.Timeseries = wr.Timeseries[:0]
+
+	// Short-circuit for no possible backfill data.
+	if !minT.Before(leaseStart) {
+		if !allowInsert {
+			wr.Timeseries = wr.Timeseries[:0]
+			log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
+		}
 		return nil
 	}
 
-	// insert allowed -> parse samples
-	acceptedMinTUnix := int64(model.TimeFromUnixNano(acceptedMinT.UnixNano()))
-	for i := range tts {
-		t := &tts[i]
+	hasBackfill, err := h.filterBackfill(wr, minT, leaseStart, clusterName, replicaName)
+	if err != nil {
+		return fmt.Errorf("could not check backfill ha lease: %#v", err)
+	}
 
-		if minTUnix < acceptedMinTUnix {
-			t.Samples = filterSamples(t.Samples, acceptedMinTUnix)
+	switch {
+	case !hasBackfill && allowInsert:
+		filterOutSampleRange(wr, minTUnix, toPromModelTime(leaseStart))
+	case hasBackfill && !allowInsert:
+		filterOutSampleRange(wr, toPromModelTime(leaseStart), maxTUnix+1) //maxTUnix+1 because filterSamples treats end as exclusive.
+	case !hasBackfill && !allowInsert:
+		wr.Timeseries = wr.Timeseries[:0]
+		log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
+	default:
```

Comment:
nit: I think this would be more clear
```suggestion
	case hasBackfill && allowInsert:
```

Reply:
Would you be OK with a comment for the default case? Saying that's the case it covers and why nothing needs to be done?

Reply:
Yes please add a comment